src/query/plan/hint_provider.hpp

```diff
@@ -0,0 +1,266 @@
+// Copyright 2023 Memgraph Ltd.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt; by using this file, you agree to be bound by the terms of the Business Source
+// License, and you may not use this file except in compliance with the Business Source License.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+/// @file
+/// This file provides textual information about possible inefficiencies in the query planner.
+/// An inefficiency is for example having a sequential scan with filtering, without the usage of indices.
+
+#pragma once
+
+#include <algorithm>
+#include <vector>
+
+#include <boost/algorithm/string.hpp>
+#include <range/v3/view.hpp>
+
+#include "query/plan/operator.hpp"
+#include "query/plan/preprocess.hpp"
+
+namespace memgraph::query::plan {
+
+std::vector<std::string> ProvidePlanHints(const LogicalOperator *plan_root, const SymbolTable &symbol_table);
+
+class PlanHintsProvider final : public HierarchicalLogicalOperatorVisitor {
```

Comment:
Why a 2nd visitor? Why not have `IndexLookupRewriter` provide the hints.

Reply:
I meant of it as a hint provider not tied directly to indices -> providing any hints possible that we might encounter in the future.